### DISCUSSION TITLE
mikro orm use insert instead of update for nested updates

### DISCUSSION THREAD

**Discussion by farid009** [mikro-orm/mikro-orm#1410] [Ideas]
**Describe the bug**
I have customer entity with one to one relation to user entity 
I use this code for updating the customer but mikro orm use insert query instead of update query for user table
```
    const customer = await this.customerRepo.findOne(id, {
      populate: ['user'],
    });


    this.customerRepo.assign(customer, { user: { name: 'farid2' } });


    await this.customerRepo.flush();

```

**To Reproduce**
just clone blow repository and start it to find a problem  for more information just read the README of project
1.https://github.com/farid009/test_mikro_orm

**Expected behavior**
use update instead of insert for user table

**Versions**

| Dependency | Version |
| - | - |
| node | v14.15.4 |
| typescript | 4.1.3 |
| mikro-orm | 4.4.1 |
| your-driver | postgresql|

**Comment by B4nan** [comment]
There are many similar issues, this is in general not supported. Your query literally means "update the user property with a new user", as any object is considered as a new entity. Assign helper is designed for updating single entity, not an entity tree. It will never produce update query for a child entity.

Changing this behaviour would be a breaking change, not a fix.

**Comment by farid009** [comment]
thank you but do you have any solution for this ?

**Comment by B4nan** [comment]
The solution is to update your user entity directly instead of trying to update the whole entity graph.

```ts
this.customerRepo.assign(customer.user, { name: 'farid2' });
// or even better
customer.user.name = 'farid2';
```

**Comment by farid009** [comment]
because I use a generic repository on top of mikro-orm I want edit the entity without knowing the structure of entity in other words without manual change is there any better solution?
also I try lodash merge function but its not work in some situation

**Comment by farid009** [comment]
I think this feature can be added to assign with a options in this case, it does not cause a break change

**Comment by B4nan** [comment]
No there is no better solution, this library is for mapping things to objects, not for doing magic :] Already saw about 10 issues asking for something similar. And each and every one had a bit different idea of how it should work... 

If you want such behaviour, write the helper method yourself, so it can fit your needs. You can leverage the ORM metadata to make it dynamic. 

> also I try lodash merge function but its not work in some situation

Using other libraries will make things even worse, as they won't respect the constraints requeired by MikroORM.

**Comment by farid009** [comment]
so many thanks 
can you please mention some related issues?
because I cant find them

**Comment by B4nan** [comment]
Most of those: https://github.com/mikro-orm/mikro-orm/issues?q=is%3Aissue+assign+nested

Also there was a recent discussion on slack about that: https://mikroorm.slack.com/archives/CFSAF8DEX/p1611249803013700

Your request is something that could live under a flag - but there are already quite a lot of flags for assign helper, which makes both the code and its usage quite complex. 

I am currently thinking about introducing new helper method that would be async and handle some of the magic, here is a summary from the slack thread:

> regarding the assign and upserts, i am still not sure how you guys expect that to work :] i gave it a shot recently, but to issue update query from a flush call, the entity needs to be loaded first. i can imagine two ways:
> 1. during the assign call, we can check whether there is a PK in given object and check the ID map to know if we should just assign on existing/already loaded entity (but that means you would first have to load entities to the context, so not very usable on its own)
> 2. make some other method, this time it would be async, and it would try to load an entity in this case, if its found, it would assign on it (and  flush would issue update query), if not, it would be registered for insert query

Still it would not work like what you tried, doing `assign(customer, { user: { name: 'farid2' } });` means "put `{ name: 'farid2' }` to the `user` property. Without providing the PK, we would still consider it as a new entity - how could we know that you want an update instead of insert? Even if we had a flag for this behaviour, I am quite sure people would want to have control over what property gets what threatment, e.g. here I want update but this one should have been insert...

**Reply by B4nan** [reply]
This discussion is more than 18 months old. You are commenting without any code examples, to the discussion that was about v4, a year after v5 was released, which changed a lot of things in terms of the `assign` helper. I don't know what you are exactly talking about, many things changed. Provide some code, so we can talk about what you actually don't like and what can we improve.

> The update or insert flag is actually the presence or absence of the id in the target object to be saved for auto-generated ids

Well, that's not how the ORM works. Your suggestion would mean it would be impossible to issue an insert query for AI PK (while providing its value). That would be a huge breaking change, and I am not interested in changing that.

> For other primary key scenarios, a select statement could be issued behind the scenes by the framework to decide if the translation should be an insert or a select.

`assign` is sync API, it can never load anything from the database.
